nexus/db-queries/src/db/datastore/volume.rs

```diff
@@ -809,6 +812,101 @@ impl DataStore {
                 public_error_from_diesel(e, ErrorHandler::Server)
             })
     }
+
+    // An Upstairs is created as part of a Volume hierarchy if the Volume
+    // Construction Request includes a "Region" variant. This may be at any
+    // layer of the Volume, and some notifications will come from an Upstairs
+    // instead of the top level of the Volume. The following functions have an
+    // Upstairs ID instead of a Volume ID for this reason.
+
+    /// Record when an Upstairs notifies us about a live repair. If that record
+    /// (uniquely identified by the four IDs passed in plus the notification
+    /// type) exists already, do nothing.
+    pub async fn live_repair_notification(
+        &self,
+        opctx: &OpContext,
+        record: LiveRepairNotification,
+    ) -> Result<(), Error> {
+        use db::schema::live_repair_notification::dsl;
+
+        let conn = self.pool_connection_authorized(opctx).await?;
+        let err = OptionalError::new();
+
+        self.transaction_retry_wrapper("live_repair_notification")
+            .transaction(&conn, |conn| {
+                let record = record.clone();
+                let err = err.clone();
+
+                async move {
+                    match &record.notification_type {
+                        LiveRepairNotificationType::Started => {
+                            // Proceed - the insertion can succeed or fail below
+                            // based on the table's primary key
+                        }
+
+                        LiveRepairNotificationType::Succeeded
+                        | LiveRepairNotificationType::Failed => {
+                            // However, Nexus must accept only one "finished"
+                            // status - an Upstairs cannot change this and must
+                            // instead perform another repair with a new repair
+                            // ID.
+                            let maybe_existing_finish_record: Option<
+                                LiveRepairNotification,
+                            > = dsl::live_repair_notification
+                                .filter(dsl::repair_id.eq(record.repair_id))
+                                .filter(dsl::upstairs_id.eq(record.upstairs_id))
+                                .filter(dsl::session_id.eq(record.session_id))
+                                .filter(dsl::region_id.eq(record.region_id))
+                                .filter(dsl::notification_type.eq_any(vec![
+                                    LiveRepairNotificationType::Succeeded,
+                                    LiveRepairNotificationType::Failed,
+                                ]))
```

Comment:
This does touch on a problem. Each Upstairs should probably retry notifying Nexus until there's a known result of that notification, like saga nodes do. Currently though the Upstairs is stateless - if it crashes, then the state of that notification retry (or any notifications) is lost. Without a place to store that state the potential for lost notifications exists.

Even without a crash it is conceivable that due to some asymmetric network outage an Upstairs would also _not_ be able to contact Nexus, but still able to receive, via propolis, volume replacement commands. It would then try to send out those notifications only to spin retrying them.

It's possible that Nexus could detect this case. I know the PR isn't up yet but the forthcoming one will contain region replacement request records. If Nexus completes the "region replacement start" saga then (ignoring the other cases for a sec) it will have sent a volume replacement request to an instance's propolis. Assuming the Propolis server stays up, there could be some action taken if Nexus does not receive any "live repair started" or "live repair finished" despite having successfully sent a volume replacement request.

There's more thinking to be done here, and probably more hairy cases to work though. For now I think it's ok to accept a Success/Failed message without a Started message to account for any transient network problems getting in the way of the Started notification.

Reply:
Yeah, I think allowing it seems like the best thing to do here.